### ISSUE TITLE
ES6 class definition & module support

### ISSUE THREAD

**Issue by obiot** [melonjs/melonJS#915]
This ticket list all the related tasks required to migrate to es6 and using a module approach. Until fully stable, the current build process will however keep transpiling everything to es5 and generate a es5 umd module. Ultimately the goal is to generate two different build :
- `melonjs.js` legacy es5 build for legacy browsers/devices (to be decided yet at which level we keep some backward compatibility with regard to the current Jay inheritance mechanism vs a standard es5 here)
- `melons-module.js` a es6 module based version for modern browser/devices (use es6 inheritance mechanism)
 
Migration phases :

* [ ] **ES6 class semantic** : to replace the current Jay inheritance mechanism (see also #685) : 
    *  [ ]  me.Object
    *  [ ]  me.Ellipse : -> me.Object
    *  [ ]  me.Pointer (used by me.input) : -> me.Object
    *  [ ]  me.Rect : -> me.Polygon
    *  [ ]  me.Polygon : -> me.Object
    *  [ ]  me.Line : -> me.Polygon
    *  [ ]  me.Renderable : -> me.Object
    *  [ ]  me.Camera2d : -> me.Renderable
    *  [ ]  me.Sprite : -> me.Renderable
    *  [ ] me.ColorLayer : -> me.Renderable
    *  [ ] me.ImageLayer : -> me.Renderable
    *  [ ] me.GUI : -> me.Renderable
    *  [ ]  me.Body : -> me.Rect
    *  [ ]  me.Entity (deprecated in version 9.0) : -> me.Renderable
    *  [ ]  me.Text : -> me.Renderable
    *  [ ]  me.BitmapText : -> me.Renderable
    *  [ ]  me.BitmapTextData (only used by me.BitmapText) : -> me.Renderable
    *  [ ]  me.Container : -> me.Renderable
    *  [ ]  me.Stage : -> me.Renderable
* [x] ~~**ES6 module export** : to allow generating a es6 bundle: 
    *  [x]  Class definition
        *  [x] me.Object
        *  [x] me.Renderable
        *  [x] me.Sprite
        *  [x] me.ColorLayer
        *  [x] me.ImageLayer
        *  [x] me.GUI
        *  [x] me.Entity (deprecated in version 9.0)
        *  [x] me.Text
        *  [x] me.BitmapText
        *  [x] me.BitmapTextData (only used by me.BitmapText)
        *  [x] me.Container
        *  [x] me.Stage
        *  [x] me.Vector2d
        *  [x] me.Vector3d
        *  [x] me.ObservableVector2d
        *  [x] me.ObservableVector3d
        *  [x] me.Matrix2d
        *  [x] me.Matrix3d
        *  [x] me.Camera2d
        *  [x] me.TMX(xxxx) 
        *  [x] me.ParticleContainer (only used by me.ParticleEmmiter)
        *  [x] me.ParticleEmmiter 
        *  [x] me.Particle
        *  [x] me.Body
        *  [x] me.Quadtree
        *  [x] me.Ellipse
        *  [x] me.Pointer (used by me.input)
        *  [x] me.Rect
        *  [x] me.Polygon
        *  [x] me.Line
        *  [x] me.Tween
        *  [x] me.Color
        *  [x] me.Renderer
        *  [x] me.Rendere.Texture
        *  [x] TextureCache (private)
        *  [x] me.CanvasRenderer
        *  [x] me.WebGLRenderer
        *  [x] me.GLShader
        *  [x] me.WebGLCompositor
    *  [x]  built-in utility class (changes to es6 will be backward compatible)
        *  [x] me 
        *  [x] me.audio
        *  [x] me.event
        *  [x] me.input
        *  [x] me.levelDirector (-> me.level)
        *  [x] me.loader
        *  [x] me.Math
        *  [x] me.collision
        *  [x] me.plugin
        *  [x] me.state
        *  [x] me.agent (-> me.utils.agent)
        *  [x] me.device
        *  [x] me.pool
        *  [x] me.save
        *  [x] me.timer
        *  [x] me.utils.*
        *  [x] me.video
        *  [x] me.game
        *  [x] ~~me.sys~~ (remove/deprecated)

* [x] ~~**ES6 module support** : with potential option to build tailored version of melonJS with only selected modules~~ -> structure and build process in place, rest will be done gradually while converting the above class definition to es6
* [x] **deprecated API** : include this automatically in the build process only once (after) the transpiling to es5
* [ ] [#678] **window.fetch** : might not be that widely supported, but a polyfill can easily be added, see https://github.com/github/fetch
* [x] ~**Babylon**~ Rollup Bubble : to automatically transpile and generate an ES5 build.

I will close all the ticket referenced in this one, just to avoid having the same topic covered in different places.

Feel free to comments here under !

**Comment by itsjavi** [comment]
Oh I am really looking forward for v6
I am currently doing nasty things in order to use Melon with Babel as you can see: https://github.com/pokettomonstaa/NewBark

Thank you for taking care of this!

**Comment by SnareChops** [comment]
@obiot How can I help move this along? I would really like to use your game engine over Phaser due to the simplicity this provides, but non es6 and non npm available makes it extremely difficult to use with typescript. I have a large amount of time I could commit to moving this project along, and only need a quick pointer to get started (ie. "Can you start converting these files" / "This is the next goal ____ start here")

**Comment by obiot** [comment]
@SnareChops Well I would looooove some help on this to be honest, I've been side tracked recently, and this is taking much longer than I initially expected...

In terms of progress though I almost converted all the classes to the ES6 syntax, what is missing is a few of them, and the whole build process that also need to be updated. So if you really have a lots of free time and want to help you can start by looking at the below branch and either finish converting the classes definition and/or update the build process + run one of the example against it :
https://github.com/melonjs/melonJS/tree/ES6-update

Can't thank you enough in advance for any help you can provide on this one :)

As for the master branch, I was planning (will do so shortly) to release an intermediate release with the current changes that includes cleaning of the API (to ensure melonJS is properly isolated within its own namesake) and all the bug fixes from the 5.x branch.

**Comment by kibertoad** [comment]
@obiot So what's the plan for the ES6 branch for now? Is it going to be synced with master at some point? Also I can't say that I have lots of free time, but I have some - what would be the top priority tasks at this point to push it all forward?

**Comment by SnareChops** [comment]
@obiot So I think I'm going to need to bow out from this one. Some of the files are very easily converted, but others where the rule of single responsibility is broken is very difficult due to the design architecture of the application. To properly convert will be a much larger undertaking than you may have initially expected due to the organizational structure that the library was written in. 

At this point I don't actually recommend converting the library to ES6, but instead create an ES6 wrapping layer over top of the ES5, then slowly convert file by file to ES6 as time goes on. Proper organization is critical for ES6 and following a strict rule of "Single Discreet Responsibility" within your files will make the transition much simpler.

I attempted to convert as much as I could and did convert every file in the project to ES6, but when trying to tie all of the imports together the problems really started to reveal themselves, and out of respect for you and your other contributors I decided not to start taking extreme measures completely reorganizing your project, though I feel thats what it really needs to make the transition properly.

By placing an ES6 wrapper over the ES5 code and reorganizing piece by piece, compiling with a babl / webpack pipeline should allow you to continue with releases while this process is going on, rather than having a branch getting stale during the conversion and then dealing with the metric ton of merge conflicts at the end attempting to pull in all of the updates from master.

If you should want or need it my fork has my completed work, but would advise against trying to use it due to the reasons that I stated above.

Best of luck with the project and I'm sorry that I wasn't able to achieve the goal, I just didn't want to cross any boundaries and disrespect you or your other contributors by reorganizing everything and making major changes to the entire project.

**Comment by scarejar** [comment]
I'm for typescript also. It catches a lot of errors, is very non-intrusive to writing javascript, and it's self documenting as you write, along with auto-complete parameters and methods. I can't convey how much typescript makes javascript so much easier to work with into words but if the melonjs team still doesn't like typescript after testing it out, I stand behind the decision!

**Comment by obiot** [comment]
@SnareChops sorry for the late reply. Yes indeed the library conversion is a long task indeed and reply rewriting/refactoring several major part of it.... Totally understand your point and definitely thank you for what you did try.

I'm interested though in your current progress, even though incomplete, It can help me progress by filling the gap with what I did so far (you can send a PR on the ES6-update branch, or send it by email to me at olivier.biot@me.com).

As for the ES6 Wrapper, it would definitely be nice to have, although I'm really wondering what is the added value at this stage, considering the current "Jay" inheritance model. And looking at my recent adventure with WeChat mini-program that requires ES6 (see [here](https://github.com/melonjs/melonJS/wiki/How-to-deploy-WeChat-Mini-Games-with-melonJS)), it kind of gave me a feeling of the limit of such a wrapper (but maybe I'm wrong).

**Comment by obiot** [comment]
@scarejar I don't have anything against Typescript actually, just I never really used it (nor needed it), and to be honest everytime someone came up with the request and a proposal for a PR it never went through. 

I agree though that TS allows to better highlight API discrepancies, and last time the topic came up, we modified/improved a few things (maybe it was with you, I do not remember).

Now if you are willing to help on this and up to task ...:P  I'll be happy to start over, implement further changes in the current API, and probably try to use something like `grunt-ts` to automatically generate the TS files ?

**Comment by SnareChops** [comment]
@obiot If you want to get started into typescript you don't need to convert anything. Just rename the files from `.js` to `.ts` and run through the typescript compiler, that's all that is needed since any valid javascript **is** valid typescript. (Typescript is an extension of the base javascript language, not a different syntax like coffee script or anything like that)

As for my work you should be able to pull it in yourself if you really want (https://github.com/SnareChops/melonJS/tree/ES6-update)

**Comment by kibertoad** [comment]
@obiot Please let me know if I could contribute to the merge somehow.

**Comment by obiot** [comment]
Thanks guys ! I just added a babel task. It's not yet automatically called during the build process, but we can enable this as soon as we start adding es6 elements.

As I was saying earlier I think that a good idea would be to progressively introduce the new syntax, maybe in the following order :
- fetch API (as listed above, and use babel to include the required polypill to XMLHttpRequest)
- ES6 New Constructor at least first to user non instantiable class (e.g. renderer, utils, etc...)
- Modules
with Babel allowing to then transpile to es5 until it's fully es6 compliant :P

**Comment by Giwayume** [comment]
`Just rename the files from .js to .ts and run through the typescript compiler, that's all that is needed since any valid javascript is valid typescript.`

This is incorrect. Typescript will complain about implicit types if you do not assign types variables. You may be able to turn that off as a compiler option. You'll also run into problems such as function parameters are assumed to be required by default. They would need to be marked as optional if they are indeed optional in the vanilla code.

Point is, typescript is a strict syntactical superset of Javascript, not an extension. It requires changes to be made to vanilla Javascript in order to compile.

**Comment by obiot** [comment]
now that basically only remains to finalise the migration to es6 inheritance for the remaining classes, closing this one and opening a new ticket : #1021
